src/main/java/appeng/fluids/helper/DualityFluidInterface.java

```diff
@@ -0,0 +1,632 @@
+				toStore.setStackSize( -toStore.getStackSize() );
+
+				// make sure strange things didn't happen...
+				final FluidStack canExtract = tank.drain( toStore.getFluidStack(), false );
+				if( canExtract == null || canExtract.amount != toStore.getStackSize() )
+				{
+					changed = true;
+				}
+				else
+				{
+					IAEFluidStack notStored = Platform.poweredInsert( src, dest, toStore, this.interfaceRequestSource );
+					toStore.setStackSize( toStore.getStackSize() - (notStored == null ? 0 : notStored.getStackSize()));
+
+					if( toStore.getStackSize() > 0 )
+					{
+						// extract items!
+						changed = true;
+						final FluidStack removed = tank.drain( toStore.getFluidStack(), true );
+						if( removed == null || toStore.getStackSize() != removed.amount )
+						{
+							throw new IllegalStateException( "bad attempt at managing tanks. ( drain )" );
+						}
+					}
+				}
+			}
+		}
+		catch( final GridAccessException e )
+		{
+			// :P
+		}
+
+		if( changed )
+		{
+			this.updatePlan( slot );
+		}
+
+		this.isWorking = -1;
+		return changed;
+	}
+
+	@Override
+	public void onFluidInventoryChanged( final IFluidHandler inventory )
+	{
+		final int slot = getTankSlot( inventory );
+		this.tankChanged[slot] = true;
+
+		if( this.isWorking == slot )
+		{
+			return;
+		}
+		
+		final boolean had = this.hasWorkToDo();
+
+		this.updatePlan( slot );
+
+		final boolean now = this.hasWorkToDo();
+
+		if( had != now )
+		{
+			try
+			{
+				if( now )
+				{
+					this.gridProxy.getTick().alertDevice( this.gridProxy.getNode() );
+				}
+				else
+				{
+					this.gridProxy.getTick().sleepDevice( this.gridProxy.getNode() );
+				}
+			}
+			catch( final GridAccessException e )
+			{
+				// :P
+			}
+		}
+	}
+
+	@Override
+	public void onChangeInventory( IItemHandler inv, int slot, InvOperation mc, ItemStack removedStack, ItemStack newStack )
+	{
+		if( this.isWorking == slot )
+		{
+			return;
+		}
+
+		if( inv == this.config )
+		{
+			this.readConfig();
+		}
+	}
+
+	public void writeToNBT( final NBTTagCompound data )
+	{
+		final NBTTagList tankContents = new NBTTagList();
+		for( int i = 0; i < NUMBER_OF_TANKS; ++i )
+		{
+			tankContents.appendTag( this.tanks[i].writeToNBT( new NBTTagCompound() ) );
+		}
+
+		data.setTag( "storage", tankContents );
+		this.config.writeToNBT( data, "config" );
+	}
+
+	public void readFromNBT( final NBTTagCompound data )
+	{
+		this.config.readFromNBT( data, "config" );
+		final NBTTagList tankContents = data.getTagList( "storage", 10 );
+		if( tankContents != null )
+		{
+			for( int i = 0; i < Math.min( NUMBER_OF_TANKS, tankContents.tagCount() ); ++i )
+			{
+				this.tanks[i].readFromNBT( tankContents.getCompoundTagAt( i ) );
+			}
+		}
+		this.readConfig();
+	}
+
+	public IItemHandler getConfig()
+	{
+		return this.config;
+	}
+	
+	public AEFluidTank getTank( final int i )
+	{
+		return this.tanks[i];
+	}
+	
+	public boolean writeTankInfo(final Map<Integer, NBTTagCompound> tagMap)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') '(' is not followed by whitespace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.whitespace.ParenPadCheck)
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') ')' is not preceded with whitespace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=checkstyle%3Acom.puppycrawl.tools.checkstyle.checks.whitespace.ParenPadCheck)